File: solutions/ruby/all-your-base/1/all_your_base.rb
class BaseConverter
  def self.convert(input_base, digits, output_base)
    raise ArgumentError if input_base < 2 || output_base < 2
    raise ArgumentError if digits.any? { |d| d < 0 || d >= input_base }

    value = digits.reduce(0) { |acc, d| acc * input_base + d }
    return [0] if value == 0

    result = []
    while value > 0
      result.unshift(value % output_base)
      value /= output_base
    end
    result
  end
end

File: solutions/python/all-your-base/3/all_your_base.py
def rebase(input_base: int, digits: list[int], output_base: int) -> list[int]:
    # Check if input base is valid
    if input_base < 2:
        raise ValueError("input base must be >= 2")
    
    # Check if output base is valid
    if output_base < 2:
        raise ValueError("output base must be >= 2")
    
    # Check if all digits are valid
    for digit in digits:
        if digit < 0 or digit >= input_base:
            raise ValueError("all digits must satisfy 0 <= d < input base")
    
    # Convert the number to base 10
    number = 0
    for digit in digits:
        number = number * input_base + digit
    
    # Handle the case where the number is 0
    if number == 0:
        return [0]
    
    # Convert the number to the output base
    result = []
    while number > 0:
        result.append(number % output_base)
        number = number // output_base
    
    # Reverse the result to get the correct order
    result.reverse()
    
    return result